### PR TITLE
merge MNIST from master

### PR DESCRIPTION
**Pull request by breznak**
+ update to choices SP/CP in MNIST

### CODE REVIEW

**Comment by breznak** [comment]
@ctrl-z-9000-times please merge this, finally uses MNIST from master in CP branch. Allows choosing SP/CP for the classification.

**Comment by ctrl-z-9000-times** [comment]
@breznak,

Thank you!  
I merge this into branch `columnPooler2`.
 * I added a flag to switch between SP & CP
 * I also found the bug and it now scores 93%, which is as high as it did before the merge.

**Comment by breznak** [comment]
Thanks! 

> I also found the bug and it now scores 93%, which is as high as it did before the merge 

out of curiosity, where was the catch?

File: src/examples/mnist/MNIST_SP.cpp [hunk 1]
```diff
@@ -24,210 +24,173 @@
...
+
+  for(auto epoch = 0u; epoch < train_dataset_iterations; epoch++) {
+    NTA_INFO << "epoch " << epoch;
     // Shuffle the training data.
-    vector<UInt> index( train_labels.size() );
-    for(auto s = 0u; s < train_labels.size(); s++)
-        index[s] = s;
-    Random(3).shuffle( index.begin(), index.end() );
+    vector<UInt> index( dataset.training_labels.size() );
+    index.assign(dataset.training_labels.cbegin(), dataset.training_labels.cend());
```
Comment (ctrl-z-9000-times) [inline src/examples/mnist/MNIST_SP.cpp]: The problem is here.  The list `index` contains an index into the training dataset.  It is shuffled before each dataset iteration.

What was:
```C++
// Create a list of the numbers [0 - trainingData.size()]
for(auto s = 0u; s < train_labels.size(); s++)
     index[s] = s;
```
Became some thing different:
```C++
// Creates a list of the data-labels for each datum in the training dataset.
index.assign(dataset.training_labels.cbegin(), dataset.training_labels.cend());
```
And it should probably changed to:
`std::iota`

Comment (breznak) [reply src/examples/mnist/MNIST_SP.cpp]: so it trained on mismatched data/labels.

Comment (ctrl-z-9000-times) [reply src/examples/mnist/MNIST_SP.cpp]: `index` should contain every index of every datum in the training dataset.

Instead it contained the numbers [0-9], which are the labels of the data, not the index of the data.
